fang/src/asynk/async_queue.rs

```diff
@@ -242,137 +285,260 @@ where
-        )
-        .await
+        let query = if backend == "PostgreSQL" {
+            REMOVE_ALL_SCHEDULED_TASK_QUERY_POSTGRES
+        } else if backend == "SQLite" {
+            REMOVE_ALL_SCHEDULED_TASK_QUERY_SQLITE
+        } else if backend == "MySQL" {
+            unimplemented!()
+        } else {
+            unreachable!()
+        };
+
+        let now_str = format!("{}", Utc::now().format("%F %T%.f+00"));
+
+        Ok(sqlx::query(query)
+            .bind(now_str)
+            .execute(transaction.acquire().await?)
+            .await?
+            .rows_affected())
     }
 
     async fn remove_task_query(
-        transaction: &mut Transaction<'_>,
-        id: Uuid,
+        transaction: &mut Transaction<'_, Any>,
+        backend: &str,
+        id: &[u8],
     ) -> Result<u64, AsyncQueueError> {
-        Self::execute_query(transaction, REMOVE_TASK_QUERY, &[&id], Some(1)).await
+        let query = if backend == "PostgreSQL" {
+            REMOVE_TASK_QUERY_POSTGRES
+        } else if backend == "SQLite" {
+            REMOVE_TASK_QUERY_SQLITE
+        } else if backend == "MySQL" {
+            unimplemented!()
+        } else {
+            unreachable!()
+        };
+
+        let result = sqlx::query(query)
+            .bind(id)
+            .execute(transaction.acquire().await?)
+            .await?
+            .rows_affected();
+
+        if result != 1 {
+            Err(AsyncQueueError::ResultError {
+                expected: 1,
+                found: result,
+            })
+        } else {
+            Ok(result)
+        }
     }
 
     async fn remove_task_by_metadata_query(
-        transaction: &mut Transaction<'_>,
+        transaction: &mut Transaction<'_, Any>,
+        backend: &str,
         task: &dyn AsyncRunnable,
     ) -> Result<u64, AsyncQueueError> {
         let metadata = serde_json::to_value(task)?;
 
         let uniq_hash = Self::calculate_hash(metadata.to_string());
 
-        Self::execute_query(
-            transaction,
-            REMOVE_TASK_BY_METADATA_QUERY,
-            &[&uniq_hash],
-            None,
-        )
-        .await
+        let query = if backend == "PostgreSQL" {
+            REMOVE_TASK_BY_METADATA_QUERY_POSTGRES
+        } else if backend == "SQLite" {
+            REMOVE_TASK_BY_METADATA_QUERY_SQLITE
+        } else if backend == "MySQL" {
+            unimplemented!()
+        } else {
+            unreachable!()
+        };
+
+        Ok(sqlx::query(query)
+            .bind(uniq_hash)
+            .execute(transaction.acquire().await?)
+            .await?
+            .rows_affected())
     }
 
     async fn remove_tasks_type_query(
-        transaction: &mut Transaction<'_>,
+        transaction: &mut Transaction<'_, Any>,
+        backend: &str,
         task_type: &str,
     ) -> Result<u64, AsyncQueueError> {
-        Self::execute_query(transaction, REMOVE_TASKS_TYPE_QUERY, &[&task_type], None).await
+        let query = if backend == "PostgreSQL" {
```

Comment:
instead of adding if cases, won't it be cleaner to create the backend trait that will implement the required queries? in this all supported backends won't have to be hardcoded in this if cases

Reply:
We are using `Any` driver from `sqlx` , so the only way i have noticed that you can get the backend is if you get a connection and call `backend_name`

that is why this [code](https://github.com/ayrat555/fang/blob/support-sqlite-mysql-asynk/fang/src/asynk/async_queue.rs#L296-L298)

[docs from sqlx](https://docs.rs/sqlx/latest/sqlx/struct.AnyConnection.html#method.backend_name)

there is no way to get the backend name without a `AnyConnection`.

Maybe we should do Connections instead of transactions ?

this way we can get backend name programatically.

Reply:
Let me re think this

Reply:
I think this will solve this. [implement backend sqlx trait that allows better code](https://github.com/ayrat555/fang/pull/141/commits/bf53cfa6af1ba4f70e113e5f20438cb8e41e9ca7)

Reply:
take a look to `connect` func if backend does not match the pool, we will provide an error.

Reply:
[better implementation with enum variants](https://github.com/ayrat555/fang/pull/141/commits/004082026c5fd9cc5e539cf05c903e357374e87f)

Reply:
I think it is better to implement it with enum variants instead of generics because there are a few databases.

Reply:
also this way Database can be inferred instead of having "trust" on user xd